In [1]:
%matplotlib inline

from pathlib import Path

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
import matplotlib.pylab as plt
from dmba import classificationSummary, gainsChart
import os

no display found. Using non-interactive Agg backend


In [2]:
# Directory

working_directory = os.getcwd()
print(working_directory)

/Users/pouyasohrab


In [3]:
# Defining the Data Frame

path = working_directory + '/dmba/UniversalBank.csv'
df = pd.read_csv(path)

In [4]:
# Understand the data

df.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [5]:
df.dtypes

ID                      int64
Age                     int64
Experience              int64
Income                  int64
ZIP Code                int64
Family                  int64
CCAvg                 float64
Education               int64
Mortgage                int64
Personal Loan           int64
Securities Account      int64
CD Account              int64
Online                  int64
CreditCard              int64
dtype: object

In [6]:
df.describe()

ID          Age   Experience       Income      ZIP Code  \
count  5000.000000  5000.000000  5000.000000  5000.000000   5000.000000   
mean   2500.500000    45.338400    20.104600    73.774200  93152.503000   
std    1443.520003    11.463166    11.467954    46.033729   2121.852197   
min       1.000000    23.000000    -3.000000     8.000000   9307.000000   
25%    1250.750000    35.000000    10.000000    39.000000  91911.000000   
50%    2500.500000    45.000000    20.000000    64.000000  93437.000000   
75%    3750.250000    55.000000    30.000000    98.000000  94608.000000   
max    5000.000000    67.000000    43.000000   224.000000  96651.000000   

            Family        CCAvg    Education     Mortgage  Personal Loan  \
count  5000.000000  5000.000000  5000.000000  5000.000000    5000.000000   
mean      2.396400     1.937938     1.881000    56.498800       0.096000   
std       1.147663     1.747659     0.839869   101.713802       0.294621   
min       1.000000     0.000000     1.000000     0.000000       0.000000   
25%       1.000000     0.700000     1.000000     0.000000       0.000000   
50%       2.000000     1.500000     2.000000     0.000000       0.000000   
75%       3.000000     2.500000     3.000000   101.000000       0.000000   
max       4.000000    10.000000     3.000000   635.000000       1.000000   

       Securities Account  CD Account       Online   CreditCard  
count         5000.000000  5000.00000  5000.000000  5000.000000  
mean             0.104400     0.06040     0.596800     0.294000  
std              0.305809     0.23825     0.490589     0.455637  
min              0.000000     0.00000     0.000000     0.000000  
25%              0.000000     0.00000     0.000000     0.000000  
50%              0.000000     0.00000     1.000000     0.000000  
75%              0.000000     0.00000     1.000000     1.000000  
max              1.000000     1.00000     1.000000     1.000000

# 8.1

In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Alter binary column variables to category types

df['Online']=df['Online'].astype('category')
df['CreditCard']=df['CreditCard'].astype('category')
df['Personal Loan']=df['Personal Loan'].astype('category')


# Practice binning numerical then categorical (won't be used in the naive bayes inputs or outputs anyways)

numerical = ['ID', 'Age', 'Experience', 'Income', 'ZIP Code', 'Family', 'CCAvg']

# Bin sizes

bin = {
    'ID': 1,
    'Age': 15,
    'Experience': 15,
    'Income': 50,
    'ZIP Code': 10000,
    'Family': 2,
    'CCAvg': 1,
}

# Binning

for col in numerical:
    bin_size = bin[col]
    df[col] = [round(val / bin_size) for val in df[col]]
    df[col] = df[col].astype('category')

# Splitting the data 60%/40%

predictors = ['Online', 'CreditCard']
outcome = ['Personal Loan']

X = pd.get_dummies(df[predictors], prefix_sep=' ', drop_first=False)
y = df[outcome]

train_df, valid_df = train_test_split(df, test_size=0.4, random_state=1)
print(train_df.shape, valid_df.shape)

(3000, 14) (2000, 14)


# Split 60%/40%

predictors = ['Online', 'CreditCard']
outcome = ['Personal Loan']

X = pd.get_dummies(df[predictors], prefix_sep=' ', drop_first=False)
y = df[outcome]

train_df, valid_df = train_test_split(df, test_size=0.4, random_state=1)
print(train_df.shape, valid_df.shape)

In [8]:
train_df.head()

ID Age Experience Income ZIP Code Family CCAvg  Education  Mortgage  \
4522  4523   2          0      1        9      0     0          1         0   
2851  2852   4          2      2       10      2     2          2         0   
2313  2314   4          2      1        9      2     0          2         0   
982    983   4          2      1        9      2     0          2         0   
1164  1165   3          1      2        9      2     4          2       327   

     Personal Loan  Securities Account  CD Account Online CreditCard  
4522             0                   0           0      0          0  
2851             0                   0           0      1          0  
2313             0                   0           0      1          1  
982              0                   0           0      0          1  
1164             1                   1           1      1          0

In [9]:
df.dtypes

ID                    category
Age                   category
Experience            category
Income                category
ZIP Code              category
Family                category
CCAvg                 category
Education                int64
Mortgage                 int64
Personal Loan         category
Securities Account       int64
CD Account               int64
Online                category
CreditCard            category
dtype: object

In [10]:
df.head()

ID Age Experience Income ZIP Code Family CCAvg  Education  Mortgage  \
0  1   2          0      1        9      2     2          1         0   
1  2   3          1      1        9      2     2          1         0   
2  3   3          1      0        9      0     1          1         0   
3  4   2          1      2        9      0     3          2         0   
4  5   2          1      1        9      2     1          2         0   

  Personal Loan  Securities Account  CD Account Online CreditCard  
0             0                   1           0      0          0  
1             0                   1           0      0          0  
2             0                   0           0      0          0  
3             0                   0           0      0          0  
4             0                   0           0      0          1

In [11]:
# 8.1.a

df_piv_data = train_df[['Online', 'CreditCard', 'Personal Loan']]
df_piv = pd.pivot_table(df_piv_data, index=['CreditCard', 'Personal Loan'], columns=['Online'], aggfunc=len, margins=True)

df_piv

Online                       0     1   All
CreditCard Personal Loan                  
0          0               792  1117  1909
           1                73   126   199
1          0               327   477   804
           1                39    49    88
All                       1231  1769  3000

In [12]:
# 8.1.b

prob_8b = ((49)/(477+49))

print('P(Loan = 1 | CC = 1 | Online = 1):', round(prob_8b*100,2), '%')

P(Loan = 1 | CC = 1 | Online = 1): 9.32 %


In [13]:
# 8.1.c

df_piv_data_2 = train_df[['Online','Personal Loan']]
df_piv_2 = pd.pivot_table(df_piv_data_2, index=['Personal Loan'], columns=['Online'], aggfunc=len, margins=True)
df_piv_2

Online            0     1   All
Personal Loan                  
0              1119  1594  2713
1               112   175   287
All            1231  1769  3000

In [14]:
# 8.1.c

df_piv_data_3 = train_df[['CreditCard','Personal Loan']]
df_piv_3 = pd.pivot_table(df_piv_data_3, index=['Personal Loan'], columns=['CreditCard'], aggfunc=len, margins=True)
df_piv_3

CreditCard        0    1   All
Personal Loan                 
0              1909  804  2713
1               199   88   287
All            2108  892  3000

In [15]:
# 8.1.d.i

prob_8d_i = ((88)/(199+88))

print('P(CC = 1 ∣ Loan = 1):', round(prob_8d_i*100,2), '%')

P(CC = 1 ∣ Loan = 1): 30.66 %


In [16]:
# 8.1.d.ii

prob_8d_ii = ((175)/(175+112))

print('P(Online = 1 ∣ Loan = 1):', round(prob_8d_ii*100,2), '%')

P(Online = 1 ∣ Loan = 1): 60.98 %


In [17]:
# 8.1.d.iii

prob_8d_iii = ((287)/(2713+287))

print('P(Loan = 1) (the proportion of loan acceptors):', round(prob_8d_iii*100,2), '%')

P(Loan = 1) (the proportion of loan acceptors): 9.57 %


In [18]:
# 8.1.d.iv

prob_8d_iv = ((804)/(1909+804))

print('P(CC = 1 ∣ Loan = 0):', round(prob_8d_iv*100,2), '%')

P(CC = 1 ∣ Loan = 0): 29.64 %


In [19]:
# 8.1.d.v

prob_8d_v = ((1594)/(1119+1594))

print('P(Online = 1 ∣ Loan = 0):', round(prob_8d_v*100,2), '%')

P(Online = 1 ∣ Loan = 0): 58.75 %


In [20]:
# 8.1.d.vi

prob_8d_vi = ((2713)/(2713+287))

print('P(Loan = 0):', round(prob_8d_vi*100,2), '%')

P(Loan = 0): 90.43 %


In [21]:
# 8.1.e

nb_calc = (prob_8d_i * prob_8d_ii * prob_8d_iii)
den = (prob_8d_i * prob_8d_ii * prob_8d_iii) + (prob_8d_iv * prob_8d_v * prob_8d_vi)
nb_calc = (nb_calc/den)
print('P(Loan = 1 | CC = 1 | Online = 1) using Naive Bayes Theorem:', round(nb_calc*100,2), '%')

P(Loan = 1 | CC = 1 | Online = 1) using Naive Bayes Theorem: 10.2 %


# 8.1.f

Naive Bayes value was slightly higher than the probability I got from using the pivot tables. Naive Bayes = 10.2% vs Pivot Table = 9.32%. I would assume that the pivot table probabilty is a more accurate estimate since the Naive Bayes estimate assumes conditional independence of all the variables. As a result, this may not always reflect accurate real-world situations.

In [22]:
# 8.1.g

# split into training and validation

X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.40, random_state=1)

In [23]:
# run naive Bayes

bnk_nb = MultinomialNB(alpha=0.01)
bnk_nb.fit(X_train, y_train)

# predict probabilities

predProb_train = bnk_nb.predict_proba(X_train)
predProb_valid = bnk_nb.predict_proba(X_valid)

# predict class membership

y_train_pred = bnk_nb.predict(X_train)
y_valid_pred = bnk_nb.predict(X_valid)

/Users/pouyasohrab/anaconda3/lib/python3.10/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [24]:
y_train_pred

array([0, 0, 0, ..., 0, 0, 0])

In [25]:
y_valid_pred

array([0, 0, 0, ..., 0, 0, 0])

In [26]:
train_df, valid_df = train_test_split(df, test_size=0.4, random_state=1)

pd.set_option('display.precision', 4)

# probability

print(train_df['Personal Loan'].value_counts() / len(train_df))
print()

for predictor in predictors:
    # construct the frequency table
    
    df = train_df[['Personal Loan', predictor]]
    freqTable = df.pivot_table(index='Personal Loan', columns=predictor,aggfunc=len)
    
    # divide each value by the sum of the row to get conditional probabilities
    
    propTable = freqTable.apply(lambda x: x / sum(x), axis=1)
    print(propTable)
    print()
pd.reset_option('display.precision')

0    0.9043
1    0.0957
Name: Personal Loan, dtype: float64

Online              0       1
Personal Loan                
0              0.4125  0.5875
1              0.3902  0.6098

CreditCard          0       1
Personal Loan                
0              0.7036  0.2964
1              0.6934  0.3066



In [27]:
# Rename variable to fix for below

valid_df.columns = valid_df.columns.str.replace('Personal Loan', 'Personal_Loan')
valid_df

ID Age Experience Income ZIP Code Family CCAvg  Education  Mortgage  \
2764  2765   2          0      2        9      0     3          3       105   
4767  4768   2          1      1        9      2     1          1       101   
3814  3815   2          1      1        9      2     1          1         0   
3499  3500   3          2      2        9      0     0          1       286   
2735  2736   2          1      1        9      2     3          2       165   
...    ...  ..        ...    ...      ...    ...   ...        ...       ...   
4372  4373   2          1      1        9      0     2          2         0   
3401  3402   3          1      1       10      0     1          3       118   
1239  1240   3          2      0        9      1     1          3       109   
437    438   2          1      1       10      2     1          2         0   
415    416   2          1      1        9      2     1          2       124   

     Personal_Loan  Securities Account  CD Account Online CreditCard  
2764             0                   0           0      0          1  
4767             0                   1           0      0          0  
3814             0                   0           0      0          0  
3499             0                   0           0      1          0  
2735             0                   0           0      1          0  
...            ...                 ...         ...    ...        ...  
4372             0                   0           0      0          0  
3401             0                   0           0      1          1  
1239             0                   0           0      1          1  
437              0                   1           0      1          0  
415              0                   0           0      1          0  

[2000 rows x 14 columns]

In [28]:
# Rename variable to fix for below

y_valid.columns = y_valid.columns.str.replace('Personal Loan', 'Personal_Loan')
y_valid

Personal_Loan
2764             0
4767             0
3814             0
3499             0
2735             0
...            ...
4372             0
3401             0
1239             0
437              0
415              0

[2000 rows x 1 columns]

In [29]:
# Code for scoring data using naive bayes

df = pd.concat([pd.DataFrame({'actual': y_valid.Personal_Loan, 'predicted': y_valid_pred}),
                pd.DataFrame(predProb_valid, index=y_valid.index)], axis=1)
mask = ((valid_df.Online == 1) & (valid_df.Personal_Loan == 1) & (valid_df.CreditCard == 1)) 


df[mask]

actual  predicted         0         1
348       1          0  0.897993  0.102007
1524      1          0  0.897993  0.102007
1273      1          0  0.897993  0.102007
1038      1          0  0.897993  0.102007
4883      1          0  0.897993  0.102007
1050      1          0  0.897993  0.102007
2809      1          0  0.897993  0.102007
891       1          0  0.897993  0.102007
566       1          0  0.897993  0.102007
2707      1          0  0.897993  0.102007
1499      1          0  0.897993  0.102007
2046      1          0  0.897993  0.102007
772       1          0  0.897993  0.102007
4927      1          0  0.897993  0.102007
4175      1          0  0.897993  0.102007
4585      1          0  0.897993  0.102007
4823      1          0  0.897993  0.102007
4147      1          0  0.897993  0.102007
438       1          0  0.897993  0.102007
3327      1          0  0.897993  0.102007
4008      1          0  0.897993  0.102007
1105      1          0  0.897993  0.102007
2443      1          0  0.897993  0.102007
1237      1          0  0.897993  0.102007
3006      1          0  0.897993  0.102007
3373      1          0  0.897993  0.102007
1672      1          0  0.897993  0.102007
75        1          0  0.897993  0.102007
47        1          0  0.897993  0.102007
1882      1          0  0.897993  0.102007
4980      1          0  0.897993  0.102007
2996      1          0  0.897993  0.102007
4705      1          0  0.897993  0.102007

# 8.1.g

Yes, the probabilities are similar to the 10.2% estimate I calculated in part 'e'. Here we see it at 10.02%.

# 8.2

In [30]:
# Defining the Data Frame

path = working_directory + '/dmba/accidentsFull.csv'
dff = pd.read_csv(path)

In [31]:
# Understand the data

dff.head()

HOUR_I_R  ALCHL_I  ALIGN_I  STRATUM_R  WRK_ZONE  WKDY_I_R  INT_HWY  \
0         0        2        2          1         0         1        0   
1         1        2        1          0         0         1        1   
2         1        2        1          0         0         1        0   
3         1        2        1          1         0         0        0   
4         1        1        1          0         0         1        0   

   LGTCON_I_R  MANCOL_I_R  PED_ACC_R  ...  SUR_COND  TRAF_CON_R  TRAF_WAY  \
0           3           0          0  ...         4           0         3   
1           3           2          0  ...         4           0         3   
2           3           2          0  ...         4           1         2   
3           3           2          0  ...         4           1         2   
4           3           2          0  ...         4           0         2   

   VEH_INVL  WEATHER_R  INJURY_CRASH  NO_INJ_I  PRPTYDMG_CRASH  FATALITIES  \
0         1          1             1         1               0           0   
1         2          2             0         0               1           0   
2         2          2             0         0               1           0   
3         2          1             0         0               1           0   
4         3          1             0         0               1           0   

   MAX_SEV_IR  
0           1  
1           0  
2           0  
3           0  
4           0  

[5 rows x 24 columns]

In [32]:
dff.dtypes

HOUR_I_R          int64
ALCHL_I           int64
ALIGN_I           int64
STRATUM_R         int64
WRK_ZONE          int64
WKDY_I_R          int64
INT_HWY           int64
LGTCON_I_R        int64
MANCOL_I_R        int64
PED_ACC_R         int64
RELJCT_I_R        int64
REL_RWY_R         int64
PROFIL_I_R        int64
SPD_LIM           int64
SUR_COND          int64
TRAF_CON_R        int64
TRAF_WAY          int64
VEH_INVL          int64
WEATHER_R         int64
INJURY_CRASH      int64
NO_INJ_I          int64
PRPTYDMG_CRASH    int64
FATALITIES        int64
MAX_SEV_IR        int64
dtype: object

In [33]:
dff.describe()

HOUR_I_R       ALCHL_I       ALIGN_I     STRATUM_R      WRK_ZONE  \
count  42183.000000  42183.000000  42183.000000  42183.000000  42183.000000   
mean       0.429344      1.912832      1.131546      0.491620      0.022616   
std        0.494988      0.282084      0.338000      0.499936      0.148677   
min        0.000000      1.000000      1.000000      0.000000      0.000000   
25%        0.000000      2.000000      1.000000      0.000000      0.000000   
50%        0.000000      2.000000      1.000000      0.000000      0.000000   
75%        1.000000      2.000000      1.000000      1.000000      0.000000   
max        1.000000      2.000000      2.000000      1.000000      1.000000   

           WKDY_I_R       INT_HWY    LGTCON_I_R    MANCOL_I_R     PED_ACC_R  \
count  42183.000000  42183.000000  42183.000000  42183.000000  42183.000000   
mean       0.771614      0.150321      1.492521      1.337079      0.040514   
std        0.419797      0.418952      0.789874      0.929756      0.197164   
min        0.000000      0.000000      1.000000      0.000000      0.000000   
25%        1.000000      0.000000      1.000000      0.000000      0.000000   
50%        1.000000      0.000000      1.000000      2.000000      0.000000   
75%        1.000000      0.000000      2.000000      2.000000      0.000000   
max        1.000000      9.000000      3.000000      2.000000      1.000000   

       ...      SUR_COND    TRAF_CON_R      TRAF_WAY      VEH_INVL  \
count  ...  42183.000000  42183.000000  42183.000000  42183.000000   
mean   ...      1.290710      0.516322      1.477491      1.816964   
std    ...      0.780524      0.749417      0.584851      0.684843   
min    ...      1.000000      0.000000      1.000000      1.000000   
25%    ...      1.000000      0.000000      1.000000      1.000000   
50%    ...      1.000000      0.000000      1.000000      2.000000   
75%    ...      1.000000      1.000000      2.000000      2.000000   
max    ...      9.000000      2.000000      3.000000     23.000000   

          WEATHER_R  INJURY_CRASH      NO_INJ_I  PRPTYDMG_CRASH    FATALITIES  \
count  42183.000000  42183.000000  42183.000000    42183.000000  42183.000000   
mean       1.142783      0.497736      0.778702        0.491217      0.011047   
std        0.349855      0.500001      1.035169        0.499929      0.104524   
min        1.000000      0.000000      0.000000        0.000000      0.000000   
25%        1.000000      0.000000      0.000000        0.000000      0.000000   
50%        1.000000      0.000000      1.000000        0.000000      0.000000   
75%        1.000000      1.000000      1.000000        1.000000      0.000000   
max        2.000000      1.000000     31.000000        1.000000      1.000000   

         MAX_SEV_IR  
count  42183.000000  
mean       0.519830  
std        0.521256  
min        0.000000  
25%        0.000000  
50%        1.000000  
75%        1.000000  
max        2.000000  

[8 rows x 24 columns]

In [34]:
# Change INJURY variable to 'yes', 'no'

import numpy as np

dff['INJURY'] = np.where((dff['MAX_SEV_IR'] == 1) | (dff['MAX_SEV_IR'] == 2), 'yes', 'no')

dff.head()


HOUR_I_R  ALCHL_I  ALIGN_I  STRATUM_R  WRK_ZONE  WKDY_I_R  INT_HWY  \
0         0        2        2          1         0         1        0   
1         1        2        1          0         0         1        1   
2         1        2        1          0         0         1        0   
3         1        2        1          1         0         0        0   
4         1        1        1          0         0         1        0   

   LGTCON_I_R  MANCOL_I_R  PED_ACC_R  ...  TRAF_CON_R  TRAF_WAY  VEH_INVL  \
0           3           0          0  ...           0         3         1   
1           3           2          0  ...           0         3         2   
2           3           2          0  ...           1         2         2   
3           3           2          0  ...           1         2         2   
4           3           2          0  ...           0         2         3   

   WEATHER_R  INJURY_CRASH  NO_INJ_I  PRPTYDMG_CRASH  FATALITIES  MAX_SEV_IR  \
0          1             1         1               0           0           1   
1          2             0         0               1           0           0   
2          2             0         0               1           0           0   
3          1             0         0               1           0           0   
4          1             0         0               1           0           0   

   INJURY  
0     yes  
1      no  
2      no  
3      no  
4      no  

[5 rows x 25 columns]

In [35]:
# 8.2.a

yes_counts = dff['INJURY'].value_counts()['yes']
total_count = dff.shape[0]

print('P(Injury = Yes):', round(yes_counts/total_count*100,2), '%')

P(Injury = Yes): 50.88 %


In [36]:
# Creating a new dataframe for 8.2.b

dff_12 = dff.head(12)

In [37]:
dff_12 = dff_12[['INJURY', 'WEATHER_R', 'TRAF_CON_R']]
dff_12

INJURY  WEATHER_R  TRAF_CON_R
0     yes          1           0
1      no          2           0
2      no          2           1
3      no          1           1
4      no          1           0
5     yes          2           0
6      no          2           0
7     yes          1           0
8      no          2           0
9      no          2           0
10     no          2           0
11     no          1           2

In [38]:
# 8.2.b.i

dff_piv = pd.pivot_table(dff_12, index=['INJURY'], columns=['WEATHER_R', 'TRAF_CON_R'], aggfunc=len, margins=True)

# Replace the NA values in the table

dff_piv.fillna(0, inplace=True)

dff_piv

WEATHER_R     1              2      All
TRAF_CON_R    0    1    2    0    1    
INJURY                                 
no          1.0  1.0  1.0  5.0  1.0   9
yes         2.0  0.0  0.0  1.0  0.0   3
All         3.0  1.0  1.0  6.0  1.0  12

In [39]:
# 8.2.b.i

dff_piv = pd.pivot_table(dff_12, index=['WEATHER_R','TRAF_CON_R'], columns=['INJURY'], aggfunc=len, margins=True)

# Replace the NA values in the table

dff_piv.fillna(0, inplace=True)

dff_piv

INJURY                 no  yes  All
WEATHER_R TRAF_CON_R               
1         0           1.0  2.0    3
          1           1.0  0.0    1
          2           1.0  0.0    1
2         0           5.0  1.0    6
          1           1.0  0.0    1
All                   9.0  3.0   12

In [40]:
# 8.2.b.ii

# 1

prob_8b_ii1 = (((2)/(3)*(3/12))/(3/12))

print('P(Injury = yes ∣ Weather_R = 1 | Traf_Con_R = 0):', 
      round(prob_8b_ii1*100,2), '%')

P(Injury = yes ∣ Weather_R = 1 | Traf_Con_R = 0): 66.67 %


In [41]:
# 8.2.b.ii2

# 2

prob_8b_ii2 = (((0)*(3/12))/(1/12))

print('P(Injury = yes ∣ Weather_R = 1 | Traf_Con_R = 1):', 
      round(prob_8b_ii2*100,2), '%')

P(Injury = yes ∣ Weather_R = 1 | Traf_Con_R = 1): 0.0 %


In [42]:
# 8.2.b.ii3

# 3

prob_8b_ii3 = (((0)*(3/12))/(1/12))

print('P(Injury = yes ∣ Weather_R = 1 | Traf_Con_R = 2):', 
      round(prob_8b_ii3*100,2), '%')

P(Injury = yes ∣ Weather_R = 1 | Traf_Con_R = 2): 0.0 %


In [43]:
# 8.2.b.ii4

# 4

prob_8b_ii4 = (((1)/(3)*(3/12))/(6/12))

print('P(Injury = yes ∣ Weather_R = 2 | Traf_Con_R = 0):', 
      round(prob_8b_ii4*100,2), '%')

P(Injury = yes ∣ Weather_R = 2 | Traf_Con_R = 0): 16.67 %


In [44]:
# 8.2.b.ii5

# 5

prob_8b_ii5 = (((0)*(3/12))/(1/12))

print('P(Injury = yes ∣ Weather_R = 2 | Traf_Con_R = 1):', 
      round(prob_8b_ii5*100,2), '%')

P(Injury = yes ∣ Weather_R = 2 | Traf_Con_R = 1): 0.0 %


In [45]:
# 8.2.b.ii6

# 6

prob_8b_ii6 = '(((0)*(3/12))/(0))'

print('P(Injury = yes ∣ Weather_R = 2 | Traf_Con_R = 2): Not possible because it is a zero in the denominator.')

P(Injury = yes ∣ Weather_R = 2 | Traf_Con_R = 2): Not possible because it is a zero in the denominator.


In [46]:
# 8.2.b.iii

import pandas as pd
import numpy as np

# Convert to a list

probabilities = [prob_8b_ii1, prob_8b_ii2, prob_8b_ii3, prob_8b_ii4, prob_8b_ii5]

# Create the DataFrame with probability variables

df = pd.DataFrame({'Probability': probabilities})

# Cutoff of 0.5
df['Injury Prediction'] = np.where(df['Probability'] >= 0.5, 'yes', 'no')

print(df)

   Probability Injury Prediction
0     0.666667               yes
1     0.000000                no
2     0.000000                no
3     0.166667                no
4     0.000000                no


In [47]:
# 8.2.b.iv

# Manual Naive Bayes

prob_8b_iv = (((2/3)*(0/3))/(3/12))

print('Naive Bayes: P(Injury = yes ∣ Weather_R = 1 | Traf_Con_R = 1):', 
      round(prob_8b_iv*100,2), '%')

Naive Bayes: P(Injury = yes ∣ Weather_R = 1 | Traf_Con_R = 1): 0.0 %


In [48]:
# Recheck dataframe

dff_12

INJURY  WEATHER_R  TRAF_CON_R
0     yes          1           0
1      no          2           0
2      no          2           1
3      no          1           1
4      no          1           0
5     yes          2           0
6      no          2           0
7     yes          1           0
8      no          2           0
9      no          2           0
10     no          2           0
11     no          1           2

In [49]:
# 8.2.b.v

import pandas as pd
from sklearn.naive_bayes import MultinomialNB

# Define predictors and outcome
predictors = ['WEATHER_R', 'TRAF_CON_R']
outcome = 'INJURY'

X = dff_12[predictors].astype('category')
y = dff_12[outcome].astype('category')

# Run naive Bayes

acc_nb = MultinomialNB(alpha=0.01)
acc_nb.fit(X, y)

# Predict probabilities

predProb = acc_nb.predict_proba(X)

# Predict class

y_pred = acc_nb.predict(X)

# Display the results

print(pd.DataFrame(predProb, columns=acc_nb.classes_))
print("\nPredicted Class:")
print(y_pred)

          no       yes
0   0.703564  0.296436
1   0.652500  0.347500
2   0.993756  0.006244
3   0.995053  0.004947
4   0.703564  0.296436
5   0.652500  0.347500
6   0.652500  0.347500
7   0.703564  0.296436
8   0.652500  0.347500
9   0.652500  0.347500
10  0.652500  0.347500
11  0.999941  0.000059

Predicted Class:
['no' 'no' 'no' 'no' 'no' 'no' 'no' 'no' 'no' 'no' 'no' 'no']


In [50]:
# Code for scoring data using naive bayes

df = pd.concat([pd.DataFrame({'actual': y, 'predicted': y_pred}),
                pd.DataFrame(predProb, index=y.index)], axis=1)
df

actual predicted         0         1
0     yes        no  0.703564  0.296436
1      no        no  0.652500  0.347500
2      no        no  0.993756  0.006244
3      no        no  0.995053  0.004947
4      no        no  0.703564  0.296436
5     yes        no  0.652500  0.347500
6      no        no  0.652500  0.347500
7     yes        no  0.703564  0.296436
8      no        no  0.652500  0.347500
9      no        no  0.652500  0.347500
10     no        no  0.652500  0.347500
11     no        no  0.999941  0.000059

# 8.2.b.v

in this model we got none of the injury = yes estimates correct. However, in the previous one completed in part 8.2.b.iii we predicted a few more correctly.

# 8.2.c.i

Predictors that should be used considering we have no information or initial reports about the accident itself are available at the time of prediction (only location characteristics, weather conditions, etc.).

Then we will include the following predictors:

TRAF_CON_R, WRK_ZONE, LGTCON_I_R, WKDY_I_R, INT_HWY, WEATHER_R, SPD_LIM, TRAF_WAY, ALIGN_I, PROFIL_I_R, HOUR_I_R, SUR_CON



In [51]:
# Check dtypes

dff.dtypes

HOUR_I_R           int64
ALCHL_I            int64
ALIGN_I            int64
STRATUM_R          int64
WRK_ZONE           int64
WKDY_I_R           int64
INT_HWY            int64
LGTCON_I_R         int64
MANCOL_I_R         int64
PED_ACC_R          int64
RELJCT_I_R         int64
REL_RWY_R          int64
PROFIL_I_R         int64
SPD_LIM            int64
SUR_COND           int64
TRAF_CON_R         int64
TRAF_WAY           int64
VEH_INVL           int64
WEATHER_R          int64
INJURY_CRASH       int64
NO_INJ_I           int64
PRPTYDMG_CRASH     int64
FATALITIES         int64
MAX_SEV_IR         int64
INJURY            object
dtype: object

In [52]:
# 8.2.c.ii

import pandas as pd
from sklearn.model_selection import train_test_split

predictors = ['TRAF_CON_R', 'WRK_ZONE', 'LGTCON_I_R', 'WKDY_I_R', 'INT_HWY', 'WEATHER_R', 'SPD_LIM', 'TRAF_WAY', 'ALIGN_I', 'PROFIL_I_R', 'HOUR_I_R', 'SUR_COND']
outcome = ['INJURY']

X = pd.get_dummies(dff[predictors].astype('category'), prefix_sep='_', drop_first=False)
y = dff[outcome]['INJURY'].astype('category')  # Convert 'INJURY' to categorical data type

# Get the list of unique classes from the categorical variable y

classes = list(y.cat.categories)

# Split 60%/40%

X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.40, random_state=1)
print(X_train.shape, X_valid.shape)

(25309, 44) (16874, 44)


In [53]:
# run naive Bayes

acc2_nb = MultinomialNB(alpha=0.01)
acc2_nb.fit(X_train, y_train)

# predict probabilities

predProb_train = acc2_nb.predict_proba(X_train)
predProb_valid = acc2_nb.predict_proba(X_valid)

# predict class membership

y_train_pred = acc2_nb.predict(X_train)
y_valid_pred = acc2_nb.predict(X_valid)

In [54]:
# Training Classification Matrix

classificationSummary(y_train, y_train_pred, class_names=classes)

Confusion Matrix (Accuracy 0.5378)

       Prediction
Actual   no  yes
    no 4712 7680
   yes 4017 8900


In [55]:
# Validation Classification Matrix

classificationSummary(y_valid, y_valid_pred, class_names=classes)

Confusion Matrix (Accuracy 0.5371)

       Prediction
Actual   no  yes
    no 3107 5222
   yes 2589 5956


In [56]:
# 8.c.iii

# Recreate Confusion Matrix with Error Rate for Training

# Positive
# Negative

tp = 8900 
fp = 7680 
tn = 4712 
fn = 4017 
total = (tn+fp+fn+tp)

# Confusion Matrix

cmatrix = np.array([[tn,fp],[fn,tp]])
print("Confusion Matrix for Training")
print(cmatrix) 

# Error Rate and Print

error1 = (fp+fn)/total
print("Error % for Training: ", round(error1*100,2))

Confusion Matrix for Training
[[4712 7680]
 [4017 8900]]
Error % for Training:  46.22


In [57]:
# 8.c.iii

# Recreate Confusion Matrix with Error Rate for Validation

# Positive
# Negative

tp = 5956 
fp = 5222 
tn = 3107 
fn = 2589 
total = (tn+fp+fn+tp)

# Confusion Matrix

cmatrix = np.array([[tn,fp],[fn,tp]])
print("Confusion Matrix for Validation")
print(cmatrix) 

# Error Rate and Print

error2 = (fp+fn)/total
print("Error % for Validation: ", round(error2*100,2))

Confusion Matrix for Validation
[[3107 5222]
 [2589 5956]]
Error % for Validation:  46.29


In [58]:
# Find the difference between the training and validation set error rate

diff = error2-error1
round(diff*100,2)

0.07

# 8.c.iv

The percent improvement relative to the naive rule (using the validation set) is approximately: 0.07%.

In [59]:
# 8.c.v

train_df, valid_df = train_test_split(dff, test_size=0.4, random_state=1)

pd.set_option('display.precision', 4)

# probability of flight status
print(train_df['INJURY'].value_counts() / len(train_df))
print()

for predictor in predictors:
    # construct the frequency table
    df = train_df[['INJURY', predictor]]
    freqTable = df.pivot_table(index='INJURY', columns=predictor,aggfunc=len)
    
    # divide each value by the sum of the row to get conditional probabilities
    propTable = freqTable.apply(lambda x: x / sum(x), axis=1)
    print(propTable)
    print()
pd.reset_option('display.precision')

yes    0.5104
no     0.4896
Name: INJURY, dtype: float64

TRAF_CON_R       0       1       2
INJURY                            
no          0.6591  0.1935  0.1474
yes         0.6218  0.2161  0.1620

WRK_ZONE       0       1
INJURY                  
no        0.9773  0.0227
yes       0.9769  0.0231

LGTCON_I_R       1       2       3
INJURY                            
no          0.6899  0.1284  0.1817
yes         0.7002  0.1088  0.1911

WKDY_I_R       0       1
INJURY                  
no        0.2176  0.7824
yes       0.2396  0.7604

INT_HWY       0       1       9
INJURY                         
no       0.8505  0.1490  0.0004
yes      0.8585  0.1407  0.0009

WEATHER_R       1       2
INJURY                   
no         0.8413  0.1587
yes        0.8741  0.1259

SPD_LIM      5       10      15      20      25      30      35      40  \
INJURY                                                                    
no       0.0002  0.0006  0.0044  0.0078  0.1085  0.0875  0.1938  0.0949   

# 8.c.v

We get near a probability of zero for P(INJURY = No ∣ SPD_LIM = 5) = 0.0002 according to the conditional probabilities pivot table. 

This near 0% probability is likely due to the fact that when traveling at such a low speed it is unlikely to get an injury (although not impossible).